### PR TITLE
added code to support adding etcds to unassigned nodes in case of dis…

### PR DESCRIPTION
**Pull request by abdulrahmantkhalifa**
added code to support adding etcds to unassigned nodes in case of disatser

### CODE REVIEW

File: templates/node.zero-os/actions.py [hunk 1]
```diff
@@ -127,11 +128,8 @@ def monitor(job):
         node = Node.from_ays(service, token, timeout=15)
         node.client.testConnectionAttempts = 0
         state = node.client.ping()
-    except RuntimeError:
+    except (RuntimeError, Client.ConnectionError, redis.TimeoutError):
```
Comment (serboctor) [inline templates/node.zero-os/actions.py]: Change redis,timeouterror to client.timeouterror

Comment (abdulrahmantkhalifa) [reply templates/node.zero-os/actions.py]: we dont throw client.timeouterror, so no point in catching it , it is added in the core0 PR though

File: pyclient/zeroos/orchestrator/sal/ETCD.py [hunk 2]
```diff
@@ -104,6 +104,9 @@ def start(self):
         if not self.container.is_port_listening(int(self.serverBind.split(":")[1])):
             raise RuntimeError('Failed to start etcd server: {}'.format(self.name))
 
+    def stop(self):
+        self.container.job.kill("etcd.{}".format(self.name))
```
Comment (khaledkbadr) [inline pyclient/zeroos/orchestrator/sal/ETCD.py]: check if job was really killed

File: templates/etcd_cluster/actions.py [hunk 3]
```diff
@@ -26,8 +26,9 @@ def configure(job):
     config = get_configuration(service.aysrepo)
 
     nodes = set()
-    for node_service in service.producers['node']:
+    for node_service in service.producers['node']:            
         nodes.add(Node.from_ays(node_service, job.context['token']))
+        service.consume(node_service)
```
Comment (khaledkbadr) [inline templates/etcd_cluster/actions.py]: you can reach node service by `service.parent.producers["node"]`

File: templates/etcd_cluster/actions.py [hunk 4]
```diff
@@ -116,57 +119,95 @@ def get_baseports(job, node, baseport, nrports):
 def watchdog_handler(job):
     from zeroos.orchestrator.sal.Container import Container
     from zeroos.orchestrator.sal.Node import Node
+    from zeroos.core0.client import Client
     from zeroos.orchestrator.sal.ETCD import ETCD
     from zeroos.orchestrator.configuration import get_jwt_token
+    import redis
```
Comment (khaledkbadr) [inline templates/etcd_cluster/actions.py]: remove

File: templates/etcd_cluster/actions.py [hunk 5]
```diff
@@ -116,57 +119,95 @@ def get_baseports(job, node, baseport, nrports):
...
+    for etcd in working_etcds:
         container = etcd.parent
+        j.tools.async.wrappers.sync(etcd.executeAction('stop', context=job.context))
+        j.tools.async.wrappers.sync(etcd.delete())
         j.tools.async.wrappers.sync(container.executeAction('stop', context=job.context))
         j.tools.async.wrappers.sync(container.delete())
 
+    # check if nodes are more than the min number for cluster deployment which is 3.
+    all_nodes = set([service.name for service in service.aysrepo.servicesFind(role='node')])
+    if len(working_model_nodes) > 3:
```
Comment (khaledkbadr) [inline templates/etcd_cluster/actions.py]: make it 1 as etcd cluster can be of 1 etcd, deploy of the current max cluster size

File: templates/etcd_cluster/actions.py [hunk 6]
```diff
@@ -116,57 +119,95 @@ def get_baseports(job, node, baseport, nrports):
...
-    for etcd in etcds:
+    for etcd in working_etcds:
         container = etcd.parent
+        j.tools.async.wrappers.sync(etcd.executeAction('stop', context=job.context))
+        j.tools.async.wrappers.sync(etcd.delete())
         j.tools.async.wrappers.sync(container.executeAction('stop', context=job.context))
         j.tools.async.wrappers.sync(container.delete())
 
+    # check if nodes are more than the min number for cluster deployment which is 3.
+    all_nodes = set([service.name for service in service.aysrepo.servicesFind(role='node')])
```
Comment (khaledkbadr) [inline templates/etcd_cluster/actions.py]: check if extra nodes are halted

File: templates/etcd_cluster/actions.py [hunk 7]
```diff
@@ -116,57 +119,95 @@ def get_baseports(job, node, baseport, nrports):
...
+    if len(working_model_nodes) > 3:
+        service.model.data.nodes = [node.name for node in working_model_nodes]
+    else:
+        service.model.data.nodes = list(all_nodes - dead_nodes)
+
+
+    # remove old nodes and etcds from producers
+    for etcd_service in service.producers['etcd']:
```
Comment (khaledkbadr) [inline templates/etcd_cluster/actions.py]: move code to where you delete etcd services for readability

File: templates/node.zero-os/actions.py [hunk 8]
```diff
@@ -390,7 +399,7 @@ def watchdog(job):
                 monitor(job)
                 cl = None
                 subscribed = None
-            except RuntimeError as e:
+            except Client.ConnectionError as e:
```
Comment (khaledkbadr) [inline templates/node.zero-os/actions.py]: move back to RuntimeError

File: pyclient/zeroos/orchestrator/sal/ETCD.py [hunk 9]
```diff
@@ -104,6 +104,15 @@ def start(self):
...
             raise RuntimeError('Failed to start etcd server: {}'.format(self.name))
 
+    def stop(self):
+        jobID = "etcd.{}".format(self.name)
+        self.container.job.kill(jobID)
+        try:
+            self.container.job.list(jobID)
+            raise RuntimeError('failed to stop etcd.')
+        except RuntimeError:
+            return
```
Comment (khaledkbadr) [inline pyclient/zeroos/orchestrator/sal/ETCD.py]: Will that code raise the error with this `except` in place?

File: templates/etcd_cluster/actions.py [hunk 10]
```diff
@@ -125,48 +128,88 @@ def watchdog_handler(job):
...
+        service.saveAll()
+    
+    for node_service in service.producers['node']:
+        if node_service.name not in service.model.data.nodes:
+            service.model.producerRemove(node_service)
+            service.saveAll()
+
+    # check if nodes are more than the min number for cluster deployment which is 3.
+    tmp = list()
+    for node in [service for service in service.aysrepo.servicesFind(role='node')]:
```
Comment (khaledkbadr) [inline templates/etcd_cluster/actions.py]: Check if the node is halted  like this `[node for node in service.aysrepo.servicesFind(role="node") if node.model.data.status != "halted"]`

Comment (abdulrahmantkhalifa) [reply templates/etcd_cluster/actions.py]: line too long that is why i didnt do that

**Comment by abdulrahmantkhalifa** [comment]
dont merge without https://github.com/zero-os/0-core/pull/489

**Pull request merged**